src/sentry/static/sentry/app/views/eventsV2/landing.tsx

```diff
@@ -58,35 +60,62 @@ class DiscoverLanding extends AsyncComponent<Props, State> {
     router: PropTypes.object.isRequired,
   };
 
-  state = {
+  state: State = {
+    // AsyncComponent state
     loading: true,
     reloading: false,
     error: false,
     errors: [],
+
+    // local component state
     isBannerHidden: checkIsBannerHidden(),
     savedQueries: [],
     savedQueriesPageLinks: '',
   };
 
   shouldReload = true;
 
+  getSavedQuerySearchQuery(): string {
+    const {location} = this.props;
+
+    return String(decodeScalar(location.query.query) || '').trim();
+  }
+
   getEndpoints(): [string, string, any][] {
     const {organization, location} = this.props;
+
     const views = getPrebuiltQueries(organization);
-    const cursor = location.query.cursor;
-    // XXX(mark) Pagination here is a bit wonky as we include the pre-built queries
-    // on the first page and aim to always have 9 results showing. If there are more than
-    // 9 pre-built queries we'll have more results on the first page. Furthermore, going
-    // back and forth between the first and second page is non-determinsitic due to the shifting
-    // per_page value.
+    const searchQuery = this.getSavedQuerySearchQuery();
+
+    const cursor = decodeScalar(location.query.cursor);
     let perPage = 9;
     if (!cursor) {
-      perPage = Math.max(1, perPage - views.length);
+      // invariant: we're on the first page
+
+      if (searchQuery && searchQuery.length > 0) {
+        const needleSearch = searchQuery.toLowerCase();
+
+        const numOfPrebuiltQueries = views.reduce((sum, view) => {
+          const eventView = EventView.fromSavedQueryWithLocation(view, location);
+
+          // if a search is performed on the list of queries, we filter
+          // on the pre-built queries
+          if (eventView.name && eventView.name.toLowerCase().includes(needleSearch)) {
+            return sum + 1;
+          }
+
+          return sum;
+        }, 0);
+
+        perPage = Math.max(1, perPage - numOfPrebuiltQueries);
+      } else {
+        perPage = Math.max(1, perPage - views.length);
+      }
     }
 
     const queryParams = {
       cursor,
-      query: 'version:2',
+      query: `version:2 ${searchQuery}`,
```

Comment:
Should this be `name:${searchQuery}` to be more explicit?

Reply:
would `name:"${searchQuery}"` be better with double quotes escaped in `searchQuery`?

Reply:
double quotes and escaping would be 💅

Reply:
Looks like double quote escaping may not be a thing since there's no lookahead performed here https://github.com/getsentry/sentry/blob/89b39704d7981ddece5ff656b7a9c094b05eba5f/src/sentry/search/utils.py#L362